### PR TITLE
feature: add docs of devops pipeline

### PR DESCRIPTION
**Pull request by shaowenchen**
/hold

### CODE REVIEW

**Comment by shaowenchen** [comment]
/unhold
/assign @FeynmanZhou 
Please review the preview directly, confirm there is no problem with the basic content.  Now there is almost no docs for DevOps Pipeline.

**Comment by FeynmanZhou** [comment]
> /unhold
> /assign @FeynmanZhou
> Please review the preview directly, confirm there is no problem with the basic content. Now there are almost no docs for DevOps Pipeline.

Got it, we will start to review and test it.

File: content/en/docs/devops-user-guide/examples/a-maven-project.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+title: "How to compile and deploy a maven project"
+keywords: 'kubernetes, docker, devops, jenkins, maven'
+description: ''
+linkTitle: "How to compile and deploy a maven project" 
+weight: 200
+---
+
+
+A maven project
```
Comment (FeynmanZhou) [inline content/en/docs/devops-user-guide/examples/a-maven-project.md:10]: Is this guide still in WIP?

File: content/en/docs/devops-user-guide/how-to-use/set-ci-node.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+description: 'How to Set CI Node for dependency cache of KubeSphere pipeline '
+---
+
+## Introduction
+
+Generally, applications often need to pull a lot of dependencies during the build process. It might cause some issues like long pulling time, or unstable network causing failure. In order to make build robust, and to speed up the build by using cache, we recommend you configure one or a set of CI nodes which the system schedules the task of CI/CD pipelines or S2I/B2I builds running on.
+
+## Label CI Nodes
+
+1. Log in KubeSphere with `admin` account, navigate to **Platform → Infrastructure**.
```
Comment (FeynmanZhou) [inline content/en/docs/devops-user-guide/how-to-use/set-ci-node.md:13]: This step should be updated and synchronized to v3.0.0, see the preview page: https://deploy-preview-367--kubesphere-v3.netlify.app/docs/devops-user-guide/how-to-use/set-ci-node/#label-ci-nodes

File: content/en/docs/devops-user-guide/how-to-use/jenkins-email.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+keywords: 'kubesphere, kubernetes, notification, jenkins, devops, ci/cd, pipeline'
+description: 'Set Email Server for KubeSphere CI/CD pipeline'
+---
+
+
+The built-in Jenkins cannot share the same email configuration with platform notification system. Thus we need to set email server for KubeSphere DevOps pipeline separately.
+
+> Note: Please be aware that the modification of the email server in `ks-jenkins` deployment below will restart the deployment itself. Consequently, the DevOps system will be unavailable for a few minutes. Please make such modification at an appropriate time.
+
+1. Log in KubeSphere by using `admin` account, navigate to **System-workspace → Projects → kubesphere-devops-system**.
```
Comment (FeynmanZhou) [inline content/en/docs/devops-user-guide/how-to-use/jenkins-email.md:12]: This step should be updated and synchronized to v3.0.0, see the preview page
https://deploy-preview-367--kubesphere-v3.netlify.app/docs/devops-user-guide/how-to-use/jenkins-email/

File: content/en/docs/devops-user-guide/how-to-use/create-a-pipeline-using-jenkinsfile.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+1.1. Log in KubeSphere with the account `project-regular`, enter into the created DevOps project and create the following three credentials under **Project Management → Credentials**:
+
+|Credential ID| Type | Where to use |
+| --- | --- | --- |
+| dockerhub-id | Account Credentials | DockerHub |
+| github-id | Account Credentials | GitHub |
+| demo-kubeconfig | kubeconfig | Kubernetes |
+
+1.2. We need to create an additional credential `sonar-token` for SonarQube token, which is used in stage 3 (SonarQube analysis) mentioned above. Refer to [Access SonarQube Console and Create Token](../../how-to-integrate/sonarqube/) to copy the token and paste here. Then press **OK** button.
```
Comment (Sherlock113) [inline content/en/docs/devops-user-guide/how-to-use/create-a-pipeline-using-jenkinsfile.md:55]: The expression "copy the token and paste here" is not clear. I noticed this guide includes "Create SonarQube Admin Token" and "Create SonarQube Token For New Project". "the token" is not clear to tell readers which one you really mean. Please adjust this part.

**Review by Sherlock113 — COMMENTED** [review]
Please have a look at my comments. I think some steps are blocked as I follow the guides. Please make changes to based on the comments. Thanks.

**Comment by FeynmanZhou** [comment]
@shaowenchen  Please take a look at our comments, update the out-of-date steps and screenshots in case mislead the users.

File: content/en/docs/devops-user-guide/examples/a-maven-project.md (hunk 1 above)
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/examples/a-maven-project.md]: I have removed it.

**Comment by shaowenchen** [comment]
> @shaowenchen Please take a look at our comments, update the out-of-date steps and screenshots in case mislead the users.

I will fix it. It will take some time.

File: content/en/docs/devops-user-guide/how-to-use/set-ci-node.md (hunk 2 above)
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/set-ci-node.md:13]: fixed

File: content/en/docs/devops-user-guide/how-to-use/jenkins-email.md (hunk 3 above)
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/jenkins-email.md]: fixed

File: content/en/docs/devops-user-guide/how-to-integrate/sonarqube.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+## Configuration of Sonarqube Server
+
+### Access SonarQube Console
+
+Now access SonarQube console `http://{$Node IP}:{$NodePort}` in your browser using the default account `admin / admin`.
+
+![](https://pek3b.qingstor.com/kubesphere-docs/png/20200107003216.png)
+
+![](https://pek3b.qingstor.com/kubesphere-docs/png/20200107003240.png)
+
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-integrate/sonarqube.md:44]: fixed

File: content/en/docs/devops-user-guide/how-to-use/devops-project-management.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
+---
+title: "DevOps Project Management"
+keywords: 'kubernetes, docker, helm, jenkins, istio, prometheus'
+description: ''
+linkTitle: "DevOps Project Management"
+weight: 110
+---
+
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/devops-project-management.md:8]: fixed

File: content/en/docs/devops-user-guide/how-to-use/create-a-pipeline-using-jenkinsfile.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+### Step 3: Create Projects
+
+In this step, we will create two projects, i.e. `kubesphere-sample-dev` and `kubesphere-sample-prod`, which are development environment and production environment respectively.
+
+#### Create The First Project
+
+> Tip：The account `project-admin` should be created in advance since it is used as the reviewer of the CI/CD Pipeline.
+
+3.1. Use the account `project-admin` to log in KubeSphere. Click **Create** button, then choose **Create a resource project**. Fill in basic information for the project. Click **Next** after complete.
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/create-a-pipeline-using-jenkinsfile.md:104]: No, I have update UI.

File: content/en/docs/devops-user-guide/how-to-use/create-a-pipeline-using-jenkinsfile.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+
+- Name: `kubesphere-sample-dev`.
+- Alias: `development environment`.
+
+
+3.2. Leave the default values at Advanced Settings. Click **Create**.
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/create-a-pipeline-using-jenkinsfile.md:110]: No, I have update UI.

File: content/en/docs/devops-user-guide/how-to-use/credential-management.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+### Create KubeConfig Credential
+
+The same as above, create a Credential, name it `demo-kubeconfig`, choose **kubeconfig**, click **Create** to complete creation.
+
+> Note: A file that is used to configure access to clusters is called a kubeconfig file. Thus we create the kubeconfig credential to access the current Kubernetes cluster, which will be used in pipeline. You don't need to change the file since KubeSphere will automatically load the kubeconfig of the current Kubernetes cluster. On the contrary, you may need to change kubeconfig in KubeSphere when access other cluster.
+
+![](https://pek3b.qingstor.com/kubesphere-docs/png/20200221174005.png)
+
+At this point, you can return to the quick start guides to complete the rest steps.
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/credential-management.md:57]: fixed

File: content/en/docs/devops-user-guide/how-to-use/jenkins-email.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+title: "Set Email Server for KubeSphere Pipeline"
+keywords: 'kubesphere, kubernetes, notification, jenkins, devops, ci/cd, pipeline'
+description: 'Set Email Server for KubeSphere CI/CD pipeline'
+---
+
+
+The built-in Jenkins cannot share the same email configuration with platform notification system. Thus we need to set email server for KubeSphere DevOps pipeline separately.
+
+> Note: Please be aware that the modification of the email server in `ks-jenkins` deployment below will restart the deployment itself. Consequently, the DevOps system will be unavailable for a few minutes. Please make such modification at an appropriate time.
+
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/jenkins-email.md:11]: fixed

File: content/en/docs/devops-user-guide/how-to-use/jenkins-email.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+1. Log in KubeSphere by using `admin` account, navigate to **System-workspace → Projects → kubesphere-devops-system**.
+
+![System Projects](https://pek3b.qingstor.com/kubesphere-docs/png/20200222231148.png)
+
+2. Then go to **Application Workloads → Workloads**, choose **Deployments** and drill into **ks-jenkins**. Then choose **Edit Config Template** from **More** drop-down list. Then choose **Pod Template** in the pop-up window and click the edit icon of the container image **ks-jenkins**. Scroll down the panel you will see the following environments that you need to configure. Finally click **Confirm** to save the changes.
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/jenkins-email.md]: fixed

File: content/en/docs/devops-user-guide/how-to-use/jenkins-setting.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+KubeSphere DevOps is based on Jenkins for containerized CI/CD workflow functionality. To provide users with a schedulable Jenkins environment, KubeSphere uses **Configuration-as-Code** for Jenkins system settings, which requires the user to log in to Jenkins Dashboard and reload after KubeSphere modifies the configuration file. In the current release, Jenkins system settings are not yet available in KubeSphere, it will be supported in upcoming release.
+
+### Modify ConfigMap
+
+If you are a cluster-admin of KubeSphere and assume you need to modify the Jenkins system configuration, it is recommended that you use Configuration-as-Code (CasC) in KubeSphere. 
+
+Firstly, you need to modify `jenkins-casc-config` in KubeSphere, then log in to Jenkins Dashboard to perform **reload**. (Because system settings written directly through Jenkins Dashborad may be overwritten by CasC configuration after Jenkins rescheduling).
+
+The built-in Jenkins CasC file is stored in `/system-workspace/kubesphere-devops-system/configmaps/jenkins-casc-config/` as **ConfigMap**, as shown below, if you need to modify it, click **Edit ConfigMap**.
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/jenkins-setting.md]: fixed

File: content/en/docs/devops-user-guide/how-to-integrate/sonarqube.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+2. Enter a project name like `java-sample`，then click **Generate**.
+
+![](https://pek3b.qingstor.com/kubesphere-docs/png/20200213230427.png)
+
+3. At this point, we've got token as follows. Click **Continue**.
+
+![](https://pek3b.qingstor.com/kubesphere-docs/png/20200213231314.png)
+
+4. Choose **Java** and select `Maven` by default, please be aware that you just need to copy the highlighted serial number, and click on the **Finish this tutorial**.
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-integrate/sonarqube.md:141]: remove it

File: content/en/docs/devops-user-guide/how-to-use/create-a-pipeline-using-jenkinsfile.md (hunk 4 above)
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-use/create-a-pipeline-using-jenkinsfile.md:55]: They have different uses.

File: content/en/docs/devops-user-guide/how-to-integrate/sonarqube.md [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+```bash
+kubectl edit  cm -n kubesphere-system  ks-console-config
+```
+
+Add the following:
+
+```bash
+client:
+    devops:
+        sonarqubeURL: http://x.x.x.x:31330
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-integrate/sonarqube.md]: fixed

File: content/en/docs/devops-user-guide/how-to-integrate/sonarqube.md [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+kubectl edit cc -n kubesphere-system ks-installer
+```
+
+Add externalSonarUrl and externalSonarToken and save it.
+
+![](/images/devops/sonarqube-token-1.png)
+
+### Add Sonarqube Server to Jenkins
+
+In the background of Jenkins, you also need to add Sonarqube Server, to use the administrator's Token.
```
[the remark this answers is not in the document]
Comment (shaowenchen) [reply content/en/docs/devops-user-guide/how-to-integrate/sonarqube.md]: fixed

**Comment by shaowenchen** [comment]
I have fixed all @FeynmanZhou @Sherlock113

**Pull request merged**